### PR TITLE
  discourse: use ghostscript-fonts for letter avatars

### PR DESCRIPTION
**Pull request by ryantm**
###### Motivation for this change

The letter avatar code expects imagmagick to be able to use
Helvetica. The ghostscript fonts package includes a helvetica
substitute that discourse uses in its Docker container.

references:
- https://meta.discourse.org/t/new-letter-avatars-are-sometimes-blank/27225/6
- https://github.com/discourse/discourse_docker/commit/8ee0f7f4c8e0090f4c12ee53a70072aac875b838

This is only needed if the setting for external avatars is turned off.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/servers/web-apps/discourse/default.nix [hunk 1]
```diff
@@ -34,7 +34,7 @@ let
     gawk
 
     # Image optimization
-    imagemagick
+    imagemagickBig # imagemagickBig is used to get get ghostscript fonts, needed for local letter avatars
```
Comment (SuperSandro2000) [inline pkgs/servers/web-apps/discourse/default.nix:37]: I think we should overwrite the imagemagick input.

Comment (ryantm) [reply pkgs/servers/web-apps/discourse/default.nix]: Thanks and done, @SuperSandro2000

**Comment by 7c6f434c** [comment]
Does this mean that imageMagickBig is also used for cropping user-uploaded images? Can it be talked into cropping a Postscript file in a way that feeds arbitrary user authored Postscript to Ghostscript?

**Comment by ryantm** [comment]
> Does this mean that imageMagickBig is also used for cropping user-uploaded images? Can it be talked into cropping a Postscript file in a way that feeds arbitrary user authored Postscript to Ghostscript?

@7c6f434c I wouldn't be able to answer that without auditing the Discourse source, but I'm guessing you are asking because you expect this to be a bad security vulnerability. Maybe I should change this to just be documentation for how someone could go about enabling the local avatar letter creation, if they really want that. By default, Discourse is configured to ask the external Discourse CDN to make letter avatars for you.

**Comment by talyz** [comment]
Discourse uses this match for the file extensions of images to be optimized: `/\A(jpe?g|png|ico|gif|webp)\z/i` and doesn't seem to check the file's signature, so given an accepted file extension, you can probably feed it postscript. Not sure if this actually makes it vulnerable to postscript based attacks, though.

**Comment by 7c6f434c** [comment]
Well, depending on what «optimise» means. Current Nixpkgs `master` branch is behind on fixing https://bugs.ghostscript.com/show_bug.cgi?id=704342 and I have tried to create a Postscript file starting `xterm` in `-dSAFER` mode using that bug and rename it to `a.png` and run `imagemagickBig` from `master` as `convert a.png b.png`. Indeed, I get an XTerm.

Postscript is a full programming language, and the `-dSAFER` in Ghostscript seems to get vulnerabilities reported periodically.

**Comment by talyz** [comment]
Optimize means running a whole slew of image optimization programs on the image, like `optipng`, `pngquant` or `jpegoptim` on the image and resizing/cropping it to the required sizes using `imagemagick`. So it definitely sounds like this would enable arbitrary command execution as the `discourse` user.

**Comment by 7c6f434c** [comment]
Would it be enough to just symlink the font file somewhere and do `makeFontsConf` and use the config?

**Comment by ryantm** [comment]
> Would it be enough to just symlink the font file somewhere and do `makeFontsConf` and use the config?

Yes. I know which ruby file to change. I'll do that.

**Comment by ryantm** [comment]
This works now, but the letters are slightly offset.

**Pull request closed without merging**